README
                        XBOMB Version 2.2
                        =================

A program that looks superficially like the minesweeper program that
comes supplied with MS Windows (and is also available in other
versions). Runs under X Windows on the UNIX operating system.

Features
--------

There are a number of features that make this version different from
the others available (that I have seen).

3 Grid Tile options

        Hexagonal      - Easy (the endgame can be difficult).
        Square         - Traditional.
        Triangular     - Difficult.

3 Grid sizes

        Small          - 8x8 with 10 bombs
        Medium         - 16x16 with 40 bombs
        Large          - 30x16 with 99 bombs

Highscore table

        10 entries for the fastest times for each of the 3 levels of
        each of the 3 grid shapes.

Playing the game
----------------

Located under the tiles of the grid are a number of unexploded
bombs. The target of the game is detect the location of the bombs
given only the number of bombs neighbouring each of the locations
known to have no bomb. Locations known to contain bombs, are marked
and the game finishes when all locations are uncovered or marked.

Left mouse button   - Uncover the current location if not already done
                      or uncover all adjacent squares not marked as bombs.
Middle mouse button - Uncover all adjacent squares not marked as bombs.
Right mouse button  - Mark a square as being known to contain a bomb (toggles).

's' key             - Start a new game
'q' key             - Quit the game
'h' key             - Print the high score table for the current grid shape.
'1','2' or '3' key  - Select game level 1, 2 or 3.
'H','S' or 'T' key  - Select game type Hexagon, Square or Triangle grid.

Installation
------------

Edit the Makefile to set the paths for installation (INSTDIR), X
windows libraries (XLIB) and the compiler options (CC, CFLAGS).

make xbomb
make install

And all should be OK.

This program has been tested on the following systems:

Linux 1.x, 2.x
SunOS 4.1.x
Solaris 2.x

Author & Copyright
------------------

Andrew M. Bishop  (XBomb first written late 1994/early 1995)

email: amb@gedanken.demon.co.uk [Please put xbomb in the subject line]

This program is Copyright Andrew M. Bishop 1995-2008 and distributed under GPL.


ChangeLog
2014-08-15  Andrew M. Bishop <amb>

        Version 2.2b Released

2014-08-15 [r76]  Andrew M. Bishop <amb>

	* FILES, README, xbomb.c, xwindow.c: Updated for version 2.2b.

2014-08-15 [r74]  Andrew M. Bishop <amb>

	* xwindow.c: Fix off-by-one error in creation of graphic contexts
	  leading to writing beyond end of array [fixes Debian bug
	  #757902]..

2013-03-16 [r73]  Andrew M. Bishop <amb>

	* README, xbomb.c: Change web-page and e-mail addresses from
	  gedanken.demon.co.uk to gedanken.org.uk.

2013-03-16 [r73]  Andrew M. Bishop <amb>

	* xbomb.c, README: Change web-page and e-mail addresses from
	  gedanken.demon.co.uk to gedanken.org.uk.

2011-10-03 [r72]  Andrew M. Bishop <amb>

	* FILES (added), ANNOUNCE (removed), LSM (removed): Don't include
	  the ANNOUNCE or LSM files in future releases. Do include the
	  FILES file in subversion.

2010-12-27 [r70]  Andrew M. Bishop <amb>

	* ChangeLog (removed): Remove the ChangeLog from version control.

2010-12-24 [r67]  Andrew M. Bishop <amb>

	* ANNOUNCE (added), COPYING (added), ChangeLog (added), LSM
	  (added): Added the uncontrolled files from the xbomb 2.2a
	  release.

2010-12-24  Andrew M. Bishop <amb@gedanken.demon.co.uk>

	Changed version control environment from RCS to CVS to SVN.

2009-11-29  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

        Version 2.2a Released

2009-11-29  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* xbomb.ad:
	Add the app-defaults for the different numbers (missed in version 2.2 release).

	* xbomb.c, xwindow.c:
	Change the colour of the number 8 from white (invisible) to light green.
	Change the version number to 2.2a.

2008-01-22  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

        Version 2.2 Released

2008-01-05  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* xbomb.6:
	Change the description of the mouse buttons, add description of the X resources
	for coloured numbers and update for version 2.2.

	* README:
	Change the description of the mouse buttons and update for version 2.2.

	* xbomb.c, xbomb.h, xwindow.c:
	Allow the left mouse button to be used to clear safe areas (like the middle
	button normally does) but only when the square has already been uncovered.

	* hiscore.c:
	Change how the player's username is determined to make it more portable.

2005-10-15  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* icon.h: Remove gcc-4.0 signed/unsigned pointer warnings.

	* hiscore.c: Include string.h.

2002-11-25  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* xwindow.c: Allow a different colour for each number.

Sat Apr 10 18:30:00 1999  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

        Version 2.1a Released

Sat Apr 10 18:28:49 1999  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* xbomb.c: Changed the version number to 2.1a.

	* xwindow.c: Improved the resizing options for the triangles game.
	Renamed the Start button to Restart.

Mon Feb 23 18:49:26 1998  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* xwindow.c:
	Changed the scaling so that smaller grids are allowed to fit a 800x640 screen.

Sat Jan 24 10:40:00 1998  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

        Version 2.1 Released

Wed Jan 21 20:20:23 1998  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* xwindow.c: Make the hiscore window highlight even when not popped up.

Wed Jan 21 19:38:18 1998  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* README: Updated for version 2.1.

Wed Jan 21 19:36:21 1998  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* xwindow.c: Stop the size changing so much.

Fri Jan 16 19:50:56 1998  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* xbomb.ad: Add in the defaults for the hi-score table.

	* hiscore.c, xbomb.c, xbomb.h, xwindow.c:
	Use global variables to simplify things.

Thu Jan 15 19:55:31 1998  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* xbomb.c, xbomb.h, xwindow.c, hiscore.c: Added a high score window in.

Sun Jan 11 11:21:44 1998  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* hiscore.c: Change the meaning of the level variable.

	* xbomb.c:
	Change the meaning of the level variable and start a game when type or level is changed.

	* xwindow.c:
	Handle window resizing better and ensure that tiles line up.

Sat Jan 10 19:17:32 1998  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* xbomb.ad: Added comments.

Thu Jan  8 20:16:12 1998  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* xwindow.c, hiscore.c, xbomb.c, xbomb.h:
	Tidied the source code and added some comments about the variables used.

Wed Jan  7 21:25:11 1998  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* xwindow.c:
	Enable the close button on the window manager menu as if the quit button was pressed.

Tue Jan 30 21:54:39 1996  Andrew M. Bishop  (amb@gedanken)

	* README: Ready to be uploaded to sunsite.

Sat Jan 20 21:02:46 1996  Andrew M. Bishop  (amb@gedanken)

	* Makefile: Added a default CC and CFLAGS.

Thu Nov 16 20:16:39 1995  Andrew Bishop  (amb@dan)

	* xbomb.c: Used rand() instead of random() to work on Solaris 2.3

	* hiscore.c: Made work on Solaris 2.3

Wed Nov 15 19:26:42 1995  Andrew Bishop  (amb@dan)

	* xwindow.c: Made work on Solaris 2.3

Tue Jul 18 20:58:37 1995  Andrew Bishop  (amb@dan)

	* xbomb.6: Fixed minor man page bug.

Tue Jul 11 19:17:38 1995  Andrew Bishop  (amb@dan)

	* Makefile: Changed X386 to X11R6.

Wed May 10 19:00:00 1995  Andrew Bishop  (amb@dan)

        Version 2.0 Released

Wed May 10 18:55:55 1995  Andrew Bishop  (amb@dan)

	* xwindow.c: Put a title on the title bar with a copyright message.

Mon Mar 20 18:51:19 1995  Andrew Bishop  (amb@dan)

	* xwindow.c:
	Changed the ratio of width:height from 15:13 to 8:7, to allow finer resizing resolution on hex and tri.

Sat Feb  4 16:48:59 1995  Andrew Bishop  (amb@dan)

	* xwindow.c: Fixed the Control-Z clock goes funny bug.

Thu Feb  2 19:01:51 1995  Andrew Bishop  (amb@dan)

	* xbomb.c, xbomb.h, hiscore.c, xwindow.c:
	Fixed warning messages that appear when compiling on sun.


xbomb.c
/***************************************
  XBomb - 'Minesweeper' game - Version 2.2b.

  Main function of program - excludes all window interface parts.
  ******************/ /******************
  Written by Andrew M. Bishop

  This file Copyright 1994-2014 Andrew M. Bishop
  It may be distributed under the GNU Public License, version 2, or
  any higher version.  See section COPYING of the GNU Public license
  for conditions under which this file may be redistributed.
  ***************************************/


#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <time.h>
#include <ctype.h>

#if defined(__sun__) && !defined(__svr4__)
int    printf(const char*, ... );
int    tolower (int c);
time_t time(time_t *t);
int    rand(void);
#endif

#include "xbomb.h"

/*+ The names of the different game +*/
char *levels[NLEVELS]={"Easy","Medium","Difficult"},      /*+ levels (difficulty). +*/
     *types [NTYPES] ={"Hexagons","Squares","Triangles"}; /*+ types (grid shapes). +*/

/*+ The size of the grids +*/
int widths [NLEVELS]={ 8 ,16 ,30},  /*+ width in tiles. +*/
    heights[NLEVELS]={ 8 ,16 ,16},  /*+ height in tiles. +*/
    nbombs [NLEVELS]={10 ,40 ,99};  /*+ number of bombs. +*/

/*+ The current game status +*/
static int status=GAME_WAIT;

/*+ The current information about the grid +*/
int grid_width,                 /*+ width. +*/
    grid_height,                /*+ height. +*/
    grid_bombs,                 /*+ number of bombs. +*/
    grid_level,                 /*+ level. +*/
    grid_type;                  /*+ type. +*/

/*+ The status of the tiles in the grid. +*/
static unsigned char state[MAX_SIZE][MAX_SIZE];

/*+ The status of the game, +*/
static int n_unseen,            /*+ the number of tiles unseen. +*/
           n_think,             /*+ the number of flags put down. +*/
           ticks;               /*+ The number of milliseconds of time passed. +*/


static int count_adjacent(int x,int y,int flag);


/*++++++++++++++++++++++++++++++++++++++
  The main program.

  int argc The command line parameters.

  char** argv The command line parameters.
  ++++++++++++++++++++++++++++++++++++++*/

int main(int argc,char** argv)
{
 int level=GAME_EASY,type=GAME_SQUARE;  /* Default values. */
 int xstatus=0;
 long t;

 printf("\n"
        "XBomb Version 2.2b\n"
        "\n"
        "(c) Andrew M. Bishop 1994-2013 [http://www.gedanken.org.uk/software/xbomb/]\n"
        "\n");

 if(argc>1)
   {
    int i,j;

    for(j=0;j<NLEVELS;j++)
      {
       char levstr[3];
       sprintf(levstr,"-%1d",j+1);
       for(i=1;i<argc;i++)
          if(!strcmp(levstr,argv[i]))
             level=GAME_LEVEL+j;
      }

    for(j=0;j<NTYPES;j++)
      {
       for(i=1;i<argc;i++)
          if(argv[i][0]=='-' && !strncasecmp(types[j],argv[i]+1,strlen(argv[i])-1))
             type=GAME_TYPE+j;
      }

    for(i=1;i<argc;i++)
       if(strlen(argv[i])>2 && !strncmp("-hiscore",argv[i],strlen(argv[i])))
         {
          grid_type=type;
          PrintHighScores();
          exit(0);
         }
   }

 t=time(&t);

 srand((unsigned int)t);

 InitialiseX(&argc,argv);

 StartGame(level,type);
 ShowHighScores();

 for(;;)
   {
    xstatus=ProcessXEvents();

    if(xstatus==GAME_QUIT)
       break;

    if(xstatus>=GAME_LEVEL && xstatus<(GAME_LEVEL+NLEVELS) && xstatus!=level)
      {
       level=xstatus;
       xstatus=GAME_START;
      }

    if(xstatus>=GAME_TYPE && xstatus<(GAME_TYPE+NTYPES) && xstatus!=type)
      {
       type=xstatus;
       xstatus=GAME_START;
      }

    if(xstatus==GAME_START)
      {
       if(status==GAME_CONTINUES)
          StopGame();
       StartGame(level,type);
       ShowHighScores();
      }

    if(status==GAME_WON)
      {
       StopGame();
       AddHighScore(ticks);
       ShowHighScores();
      }
    else if(status==GAME_LOST)
      {
       StopGame();
       ShowHighScores();
      }
   } 

 FinishUpX();

 return(0);
}


/*++++++++++++++++++++++++++++++++++++++
  Start a new game.

  int level The level of the game to play.

  int type The type of game to play.
  ++++++++++++++++++++++++++++++++++++++*/

void StartGame(int level,int type)
{
 int x,y;

 grid_type=type;
 grid_level=level;
 grid_width=widths[level-GAME_LEVEL];
 grid_height=heights[level-GAME_LEVEL];
 grid_bombs=nbombs[level-GAME_LEVEL];

 n_unseen=grid_width*grid_height;
 n_think=0;

 status=GAME_READY;

 for(x=0;x<grid_width;x++)
    for(y=0;y<grid_height;y++)
       state[x][y]=UNSEEN;

 ScaleWindow();
 DrawGrid();
 StartClock(0);
 SetUXB(grid_bombs);
}


/*++++++++++++++++++++++++++++++++++++++
  Hides the bombs in the grid.

  int xs The location that has no neighbouring bombs.

  int ys The location that has no neighbouring bombs.
  ++++++++++++++++++++++++++++++++++++++*/

void HideBombs(int xs,int ys)
{
 int i,x,y;

 state[xs][ys]+=CORRECT;

 for(i=0;i<grid_bombs;i++)
   {
    do{
       x=(rand()>>8)%grid_width;
       y=(rand()>>8)%grid_height;
      }
    while(state[x][y]!=UNSEEN || count_adjacent(x,y,CORRECT));

    state[x][y]+=ACTUAL_BOMB;
   }

 state[xs][ys]-=CORRECT;

 for(x=0;x<grid_width;x++)
    for(y=0;y<grid_height;y++)
       if(!(state[x][y]&ACTUAL_BOMB))
          state[x][y]+=count_adjacent(x,y,ACTUAL_BOMB);

 StartClock(1);
}


/*++++++++++++++++++++++++++++++++++++++
  Stop the game.
  ++++++++++++++++++++++++++++++++++++++*/

void StopGame(void)
{
 status=GAME_WAIT;
 ticks=StopClock();
}


/*++++++++++++++++++++++++++++++++++++++
  Draws the grid at the start of the game or for expose events.
  ++++++++++++++++++++++++++++++++++++++*/

void DrawGrid(void)
{
 int x,y;

 for(x=0;x<grid_width;x++)
    for(y=0;y<grid_height;y++)
       DrawSquare(x,y,state[x][y]);
}


/*++++++++++++++++++++++++++++++++++++++
  Select a single square and clear it or explode the bomb.

  int x The location of the square.

  int y The location of the square.
  ++++++++++++++++++++++++++++++++++++++*/

void SelectSquare(int x,int y)
{
 if(!(state[x][y]&UNSEEN) || state[x][y]&THINK_BOMB || status==GAME_WAIT)
    return;

 if(status==GAME_READY)
   {
    HideBombs(x,y);
    status=GAME_CONTINUES;
   }

 RemoveEmpties(x,y);

 if(state[x][y]&ACTUAL_BOMB)
   {
    int x,y;

    for(x=0;x<grid_width;x++)
       for(y=0;y<grid_width;y++)
          if(state[x][y]&UNSEEN)
            {
             state[x][y]&=~UNSEEN;
             if(state[x][y]&THINK_BOMB && state[x][y]&ACTUAL_BOMB)
                state[x][y]|=CORRECT;
            }
          else
             state[x][y]|=CORRECT;

    status=GAME_LOST;
    DrawGrid();
   }
 else
   {
    DrawSquare(x,y,state[x][y]);

    if(n_think==n_unseen)
       status=GAME_WON;
   }
}


/*++++++++++++++++++++++++++++++++++++++
  Select a single square and clear it, explode the bomb or clear the adjacent.

  int x The location of the square.

  int y The location of the square.
  ++++++++++++++++++++++++++++++++++++++*/

void SelectSquareOrAdjacent(int x,int y)
{
 if(state[x][y]&UNSEEN)
    SelectSquare(x,y);
 else /* if(state[x][y]&EMPTY) */
    SelectAdjacent(x,y);
}


/*++++++++++++++++++++++++++++++++++++++
  Select all of the squares adjacent

  int x The location of the square.

  int y The location of the square.
  ++++++++++++++++++++++++++++++++++++++*/

void SelectAdjacent(int x,int y)
{
 int dx,dy,n,dd=(grid_type==GAME_TRIANGLE?2:1);

 if(state[x][y]&UNSEEN || status==GAME_WAIT)
    return;

 n=count_adjacent(x,y,THINK_BOMB);

 if(state[x][y]!=n)
    return;

 for(dx=-dd;dx<=dd;dx++)
   {
    if((x+dx)<0 || (x+dx)>=grid_width)
       continue;

    for(dy=-1;dy<=1;dy++)
      {
       if((y+dy)<0 || (y+dy)>=grid_height)
          continue;
       if(dx==0 && dy==0)
          continue;

       if(grid_type==GAME_HEXAGON && dy!=0 && dx==(1-2*(y%2)))
          continue;
       if(grid_type==GAME_TRIANGLE && dy==(2*((x+y)%2)-1) && (dx==-2 || dx==2))
          continue;

       if(!(state[x+dx][y+dy]&THINK_BOMB))
          SelectSquare(x+dx,y+dy);
      }
   }
}


/*++++++++++++++++++++++++++++++++++++++
  Marks as a bomb the square.

  int x The location of the square.

  int y The location of the square.
  ++++++++++++++++++++++++++++++++++++++*/

void MarkBomb(int x,int y)
{
 if(!(state[x][y]&UNSEEN) || status==GAME_WAIT)
    return;

 if(n_think==grid_bombs && !(state[x][y]&THINK_BOMB))
    return;

 state[x][y]^=THINK_BOMB;

 if(state[x][y]&THINK_BOMB)
    n_think++;
 else
    n_think--;

 SetUXB(grid_bombs-n_think);

 DrawSquare(x,y,state[x][y]);

 if(n_think==n_unseen)
    status=GAME_WON;
}


/*++++++++++++++++++++++++++++++++++++++
  Removes the empty squares around here

  int x Start from this position.

  int y Start from this position.
  ++++++++++++++++++++++++++++++++++++++*/

void RemoveEmpties(int x, int y)
{
 int dx,dy,dd=(grid_type==GAME_TRIANGLE?2:1);

 if(!(state[x][y]&UNSEEN) || state[x][y]&THINK_BOMB)
    return;

 state[x][y]&=~UNSEEN;
 n_unseen--;
 DrawSquare(x,y,state[x][y]);

 if(state[x][y]!=EMPTY)
    return;

 for(dx=-dd;dx<=dd;dx++)
   {
    if((x+dx)<0 || (x+dx)>=grid_width)
       continue;

    for(dy=-1;dy<=1;dy++)
      {
       if((y+dy)<0 || (y+dy)>=grid_height)
          continue;
       if(dx==0 && dy==0)
          continue;

       if(grid_type==GAME_HEXAGON && dy!=0 && dx==(1-2*(y%2)))
          continue;
       if(grid_type==GAME_TRIANGLE && dy==(2*((x+y)%2)-1) && (dx==-2 || dx==2))
          continue;

       RemoveEmpties(x+dx,y+dy);
      }
   }
}


/*++++++++++++++++++++++++++++++++++++++
  Count the number of adjacent squares that are bombs.

  int count_adjacent Returns the number of adjacent tiles with the specified flag.

  int x The position to start from.

  int y The position to start from.

  int flag The type of thing to look for.
  ++++++++++++++++++++++++++++++++++++++*/

static int count_adjacent(int x,int y,int flag)
{
 int n=0,dx,dy,dd=(grid_type==GAME_TRIANGLE?2:1);

 for(dx=-dd;dx<=dd;dx++)
   {
    if((x+dx)<0 || (x+dx)>=grid_width)
       continue;

    for(dy=-1;dy<=1;dy++)
      {
       if((y+dy)<0 || (y+dy)>=grid_height)
          continue;

       if(grid_type==GAME_HEXAGON && dy!=0 && dx==(1-2*(y%2)))
          continue;
       if(grid_type==GAME_TRIANGLE && dy==(2*((x+y)%2)-1) && (dx==-2 || dx==2))
          continue;

       if(state[x+dx][y+dy]&flag)
          n++;
      }
   }

 return(n);
}


hiscore.c
/***************************************
  $Header: /home/amb/xbomb/RCS/hiscore.c 1.13 2008/01/05 19:23:07 amb Exp $

  XBomb - 'Minesweeper' game - Version 2.2.

  Hi-score table management.
  ******************/ /******************
  Written by Andrew M. Bishop

  This file Copyright 1994-2007 Andrew M. Bishop
  It may be distributed under the GNU Public License, version 2, or
  any higher version.  See section COPYING of the GNU Public license
  for conditions under which this file may be redistributed.
  ***************************************/


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <pwd.h>

#if defined(__sun__) && !defined(__svr4__)
int    fprintf(FILE*, const char*,...);
int    printf(const char*, ... );
int    sscanf(char*, const char*,...);
int    fread(void*,unsigned int,unsigned int, FILE*);
int    fwrite(const void*,unsigned int,unsigned int, FILE*);
int    fclose(FILE*);
long   time(long*);
#endif

#include <sys/stat.h>
#include <unistd.h>
#include <time.h>

#include "xbomb.h"

/*+ The current information about the grid +*/
extern int grid_type,           /*+ type. +*/
           grid_level;          /*+ level. +*/

/*+ The names of the different game +*/
extern char *levels[NLEVELS],   /*+ levels (difficulty). +*/
            *types[NTYPES];     /*+ types (grid shapes). +*/

/*+ The size of the grids +*/
extern int widths[NLEVELS],     /*+ width in tiles. +*/
           heights[NLEVELS],    /*+ height in tiles. +*/
           nbombs[NLEVELS];     /*+ number of bombs. +*/

/*+ The names of the high score tables. +*/
static char *filenames[NTYPES]={"/var/tmp/xbomb6.hi","/var/tmp/xbomb4.hi","/var/tmp/xbomb3.hi"};

/*+ The names of the positions in the high score tables. +*/
static char pos[11][5]={"Top","2nd","3rd","4th","5th","6th","7th","8th","9th","10th","Lost"};

/*+ The list of hi-scores for one grid type. +*/
static int  score[NLEVELS+1][11];

/*+ The list of hi-score usernames for one grid type. +*/
static char name[NLEVELS+1][11][21];

/*+ The list of hi-score dates for one grid type. +*/
static long date[NLEVELS+1][11];

/*+ Which of the hoigh scores in the table to highlight. +*/
static int which_hiscore=-1;

/*+ A structure to store the high score information in the file. +*/
struct score_name
{
 long score;                    /*+ The time in milliseconds. +*/
 char name[20];                 /*+ The user name. +*/
 long date;                     /*+ The time in seconds from the epoch. +*/
};


static void load_high_scores(void);
static void save_high_scores(void);

static void decrypt_score(struct score_name* sn);
static void encrypt_score(struct score_name* sn);


/*++++++++++++++++++++++++++++++++++++++
  Prints the high score tables.
  ++++++++++++++++++++++++++++++++++++++*/

void PrintHighScores(void)
{
 int l,j;

 load_high_scores();

 printf("\nHigh score tables for %s\n",types[grid_type-GAME_TYPE]);

 for(l=0;l<NLEVELS;l++)
   {
    printf("\nHigh scores for %s level (%dx%d grid of %s with %d bombs)\n",levels[l],widths[l],heights[l],types[grid_type-GAME_TYPE],nbombs[l]);
    for(j=0;j<10;j++)
      {
       printf("%4s : %12s = ",pos[j],name[l][j]);
       if(!date[l][j])
          printf("  SLOW");
       else if(l==0)
          printf("%6.2f",(double)score[l][j]/1000.0);
       else if(l==1)
          printf("%6.1f",(double)score[l][j]/1000.0);
       else
          printf("%6.0f",(double)score[l][j]/1000.0);
       if(date[l][j])
          printf(" : %s",ctime(&date[l][j]));
       else
         {printf(" : Never\n");break;}
      }
    printf("\n");
   }
}


/*++++++++++++++++++++++++++++++++++++++
  Adds a high score to the table if it is good enough.

  int ticks The time that was taken.
  ++++++++++++++++++++++++++++++++++++++*/

void AddHighScore(int ticks)
{
 int j;
 int changed=-1;
 struct passwd *pwd;
 char *username;

 load_high_scores();

 pwd=getpwuid(geteuid());
 if(pwd)
    username=pwd->pw_name;
 else
    username="?unknown?";

 which_hiscore=10;
 score[grid_level-GAME_LEVEL][10]=ticks;
 strcpy(name[grid_level-GAME_LEVEL][10],username);
 date[grid_level-GAME_LEVEL][10]=time(NULL);

 for(j=0;j<10;j++)
    if(ticks<score[grid_level-GAME_LEVEL][j])
       {changed=which_hiscore=j;break;}

 if(changed!=-1)
   {
    for(j=9;j>changed;j--)
      {
       score[grid_level-GAME_LEVEL][j]=score[grid_level-GAME_LEVEL][j-1];
       strcpy(name[grid_level-GAME_LEVEL][j],name[grid_level-GAME_LEVEL][j-1]);
       date[grid_level-GAME_LEVEL][j]=date[grid_level-GAME_LEVEL][j-1];
      }
    score[grid_level-GAME_LEVEL][changed]=ticks;
    strcpy(name[grid_level-GAME_LEVEL][changed],username);
    date[grid_level-GAME_LEVEL][changed]=time(NULL);

    save_high_scores();
   }
}


/*++++++++++++++++++++++++++++++++++++++
  Select a set of high scores for display.
  ++++++++++++++++++++++++++++++++++++++*/

void ShowHighScores(void)
{
 char datestr[11][32],*scores[11][4],scorestr[11][10];
 int i;

 load_high_scores();

 for(i=0;i<11;i++)
   {
    if(i==10 && which_hiscore!=10)
      {scores[i][0]=scores[i][1]=scores[i][2]=scores[i][3]="";continue;}

    strcpy(datestr[i],ctime(&date[grid_level-GAME_LEVEL][i]));

    scores[i][0]=pos[i];

    scores[i][1]=name[grid_level-GAME_LEVEL][i];

    if(!date[grid_level-GAME_LEVEL][i])
       scores[i][2]="SLOW";
    else if(grid_level==GAME_EASY)
      {sprintf(scorestr[i],"%6.2f",(double)score[grid_level-GAME_LEVEL][i]/1000.0);scores[i][2]=scorestr[i];}
    else if(grid_level==GAME_MEDIUM)
      {sprintf(scorestr[i],"%6.1f",(double)score[grid_level-GAME_LEVEL][i]/1000.0);scores[i][2]=scorestr[i];}
    else
      {sprintf(scorestr[i],"%6.0f",(double)score[grid_level-GAME_LEVEL][i]/1000.0);scores[i][2]=scorestr[i];}

    if(date[grid_level-GAME_LEVEL][i])
       scores[i][3]=datestr[i];
    else
       scores[i][3]="Never";
   }

 DisplayHighScores(scores,which_hiscore);

 which_hiscore=-1;
}


/*++++++++++++++++++++++++++++++++++++++
  Load in the high score table.
  ++++++++++++++++++++++++++++++++++++++*/

static void load_high_scores(void)
{
 FILE *f;
 int i,j;
 struct score_name sn;

 for(i=0;i<NLEVELS;i++)
    for(j=0;j<10;j++)
      {
       score[i][j]=9999999;
       strcpy(name[i][j],"nobody");
       date[i][j]=0;
      }

 f=fopen(filenames[grid_type-GAME_TYPE],"r");

 if(!f)
   {
    fprintf(stderr,"XBomb: Cannot open high score table '%s' to read.\n",filenames[grid_type-GAME_TYPE]);
    return;
   }

 for(i=0;i<NLEVELS;i++)
    for(j=0;j<10;j++)
      {
       if(!fread(&sn,sizeof(sn),1,f))
          break;
       decrypt_score(&sn);
       score[i][j]=sn.score;
       strcpy(name[i][j],sn.name);
       date[i][j]=sn.date;
      }

 fclose(f);
}


/*++++++++++++++++++++++++++++++++++++++
  Saves the high score table.
  ++++++++++++++++++++++++++++++++++++++*/

static void save_high_scores(void)
{
 FILE *f;
 int i,j;
 struct score_name sn;

 f=fopen(filenames[grid_type-GAME_TYPE],"w");

 if(!f)
   {
    fprintf(stderr,"XBomb: Cannot open high score table '%s' to write.\n",filenames[grid_type-GAME_TYPE]);
    return;
   }

 for(i=0;i<NLEVELS;i++)
    for(j=0;j<10;j++)
      {
       sn.score=score[i][j];
       strcpy(sn.name,name[i][j]);
       sn.date=date[i][j];
       encrypt_score(&sn);
       fwrite(&sn,sizeof(sn),1,f);
      }

 fclose(f);

 chmod(filenames[grid_type-GAME_TYPE],S_IRUSR|S_IWUSR|S_IRGRP|S_IWGRP|S_IROTH|S_IWOTH);
}


/*++++++++++++++++++++++++++++++++++++++
  Decrypts the high score table

  struct score_name* sn The score and name to decrypt.
  ++++++++++++++++++++++++++++++++++++++*/

static void decrypt_score(struct score_name* sn)
{
 int i;
 char *p=(char*)sn;

 for(i=0;i<sizeof(struct score_name);i++)
    p[i]^=125;

 return;
}


/*++++++++++++++++++++++++++++++++++++++
  Encrypts the high score table

  struct score_name* sn The score and name to encrypt.
  ++++++++++++++++++++++++++++++++++++++*/

static void encrypt_score(struct score_name* sn)
{
 int i;
 char *p=(char*)sn;

 for(i=0;i<sizeof(struct score_name);i++)
    p[i]^=125;

 return;
}


xbomb.h
/***************************************
  $Header: /home/amb/xbomb/RCS/xbomb.h 1.9 2008/01/05 19:24:16 amb Exp $

  XBomb - 'Minesweeper' game - Version 2.2.

  Global header file.
  ******************/ /******************
  Written by Andrew M. Bishop

  This file Copyright 1994-2007 Andrew M. Bishop
  It may be distributed under the GNU Public License, version 2, or
  any higher version.  See section COPYING of the GNU Public license
  for conditions under which this file may be redistributed.
  ***************************************/


#ifndef XBOMB_H
#define XBOMB_H

/* State of each square */

#define EMPTY         0         /*+ No bomb. +*/
#define ACTUAL_BOMB  16         /*+ Really is a bomb. +*/
#define THINK_BOMB   32         /*+ User thinks there is a bomb. +*/
#define UNSEEN       64         /*+ Not seen by user. +*/
#define CORRECT     128         /*+ guessed correctly at end of game. +*/

/* Status of game */

#define GAME_CONTINUES 0        /*+ Normal state during game. +*/
#define GAME_WAIT      1        /*+ Waiting for start to be pressed.  +*/
#define GAME_LOST      2        /*+ Just finished a game and lost. +*/
#define GAME_WON       3        /*+ Just finished a game and won. +*/
#define GAME_READY     4        /*+ Ready to start playing. +*/

/* Status of game from the X part */

#define GAME_START     10       /*+ User has pressed start button. +*/
#define GAME_QUIT      11       /*+ User has pressed quit button. +*/

/* The level of difficulty */

#define GAME_LEVEL     12       /*+ Starting point for the levels. +*/
#define GAME_EASY      12       /*+ The Easy level. +*/
#define GAME_MEDIUM    13       /*+ The Medium level. +*/
#define GAME_DIFFICULT 14       /*+ The Difficult level. +*/

/* The type of game */

#define GAME_TYPE     20        /*+ Starting point for the game types. +*/
#define GAME_HEXAGON  20        /*+ The hexagons game. +*/
#define GAME_SQUARE   21        /*+ The squares game. +*/
#define GAME_TRIANGLE 22        /*+ The triangles game. +*/

/* The levels */

#define NLEVELS  3              /*+ The number of levels. +*/
#define MAX_SIZE 30             /*+ The maximum size of the grid. +*/

/* The game types */

#define NTYPES    3             /*+ The number of types of game. +*/

/* In hiscore.c */

void PrintHighScores(void);
void AddHighScore(int ticks);
void ShowHighScores(void);

/* In xbomb.c */

void StartGame(int level,int type);
void HideBombs(int xs,int ys);
void StopGame(void);
void DrawGrid(void);
void SelectSquare(int x,int y);
void SelectSquareOrAdjacent(int x,int y);
void SelectAdjacent(int x,int y);
void MarkBomb(int x,int y);
void RemoveEmpties(int x, int y);

/* In xwindow.c */

void InitialiseX(int *argc,char **argv);
void FinishUpX(void);
int ProcessXEvents(void);
void DrawSquare(int x,int y,unsigned char value);
void DisplayHighScores(char *scores[11][4],int which_score);
void ScaleWindow(void);
void StartClock(int reset);
int StopClock(void);
void SetUXB(int n);

#endif
